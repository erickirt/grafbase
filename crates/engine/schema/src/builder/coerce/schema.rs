use crate::{
    EnumDefinitionId, InputObjectDefinitionId, InputValueDefinitionId, ScalarDefinitionId, ScalarType,
    SchemaInputValueId, SchemaInputValueRecord, TypeDefinitionId, TypeRecord,
    builder::{GraphBuilder, sdl::ConstValue},
};
use id_newtypes::IdRange;
use itertools::Itertools as _;
use wrapping::{ListWrapping, MutableWrapping};

use super::{InputValueError, can_coerce_to_int};

impl GraphBuilder<'_> {
    pub(crate) fn coerce_input_value(
        &mut self,
        id: InputValueDefinitionId,
        value: ConstValue,
    ) -> Result<SchemaInputValueId, InputValueError> {
        self.value_path.clear();
        self.value_path.push(self.graph[id].name_id.into());
        let value = coerce_input_value(self, self.graph[id].ty_record, value)?;
        Ok(self.graph.input_values.push_value(value))
    }
}

fn coerce_input_value(
    builder: &mut GraphBuilder,
    ty: TypeRecord,
    value: ConstValue,
) -> Result<SchemaInputValueRecord, InputValueError> {
    if ty.wrapping.is_list() && !value.is_list() && !value.is_null() {
        let mut value = coerce_named_type_value(builder, ty.definition_id, value)?;
        for _ in 0..ty.wrapping.list_wrappings().len() {
            value = SchemaInputValueRecord::List(IdRange::from_single(builder.graph.input_values.push_value(value)));
        }
        return Ok(value);
    }

    coerce_type_value(builder, ty.definition_id, ty.wrapping.into(), value)
}

fn coerce_type_value(
    builder: &mut GraphBuilder,
    definition_id: TypeDefinitionId,
    mut wrapping: MutableWrapping,
    value: ConstValue,
) -> Result<SchemaInputValueRecord, InputValueError> {
    let Some(list_wrapping) = wrapping.pop_outermost_list_wrapping() else {
        if value.is_null() {
            if wrapping.is_required() {
                return Err(InputValueError::UnexpectedNull {
                    expected: builder.type_name(TypeRecord {
                        definition_id,
                        wrapping: wrapping.into(),
                    }),
                    path: builder.value_path_string(),
                });
            }
            return Ok(SchemaInputValueRecord::Null);
        }
        return coerce_named_type_value(builder, definition_id, value);
    };

    match (value, list_wrapping) {
        (ConstValue::Null(_), ListWrapping::ListNonNull) => Err(InputValueError::UnexpectedNull {
            expected: builder.type_name(TypeRecord {
                definition_id,
                wrapping: {
                    wrapping.push_outermost_list_wrapping(list_wrapping);
                    wrapping.into()
                },
            }),
            path: builder.value_path_string(),
        }),
        (ConstValue::Null(_), ListWrapping::List) => Ok(SchemaInputValueRecord::Null),
        (ConstValue::List(array), _) => {
            let ids = builder.graph.input_values.reserve_list(array.len());
            for ((idx, value), id) in array.into_iter().enumerate().zip(ids) {
                builder.value_path.push(idx.into());
                builder.graph.input_values[id] = coerce_type_value(builder, definition_id, wrapping.clone(), value)?;
                builder.value_path.pop();
            }
            Ok(SchemaInputValueRecord::List(ids))
        }
        (value, _) => Err(InputValueError::MissingList {
            actual: value.into(),
            expected: builder.type_name(TypeRecord {
                definition_id,
                wrapping: {
                    wrapping.push_outermost_list_wrapping(list_wrapping);
                    wrapping.into()
                },
            }),
            path: builder.value_path_string(),
        }),
    }
}

fn coerce_named_type_value(
    builder: &mut GraphBuilder,
    definition_id: TypeDefinitionId,
    value: ConstValue,
) -> Result<SchemaInputValueRecord, InputValueError> {
    match definition_id {
        TypeDefinitionId::Scalar(id) => coerce_scalar_value(builder, id, value),
        TypeDefinitionId::Enum(id) => coerce_enum_value(builder, id, value),
        TypeDefinitionId::InputObject(id) => coerce_input_object_value(builder, id, value),
        _ => unreachable!("Cannot be an output type."),
    }
}

fn coerce_input_object_value(
    builder: &mut GraphBuilder,
    input_object_id: InputObjectDefinitionId,
    obj: ConstValue,
) -> Result<SchemaInputValueRecord, InputValueError> {
    let input_object = &builder.graph[input_object_id];
    let ConstValue::Object(object) = obj else {
        return Err(InputValueError::MissingObject {
            name: builder[input_object.name_id].clone(),
            actual: obj.into(),
            path: builder.value_path_string(),
        });
    };

    let mut fields_buffer = builder.input_fields_buffer_pool.pop().unwrap_or_default();
    let mut fields = object.fields().collect::<Vec<_>>();

    if input_object.is_one_of {
        if fields.len() != 1 {
            return Err(InputValueError::ExactlyOneFIeldMustBePresentForOneOfInputObjects {
                name: builder[builder.graph[input_object_id].name_id].clone(),
                message: if fields.is_empty() {
                    "No field was provided".to_string()
                } else {
                    format!(
                        "{} fields ({}) were provided",
                        fields.len(),
                        fields
                            .iter()
                            .format_with(",", |field, f| f(&format_args!("{}", field.name())))
                    )
                },
                path: builder.value_path_string(),
            });
        }
        let name = fields[0].name();
        if let Some(id) = input_object
            .input_field_ids
            .into_iter()
            .find(|id| builder[builder.graph[*id].name_id] == name)
        {
            let input_field = &builder.graph[id];
            let field = fields.pop().unwrap();
            builder.value_path.push(input_field.name_id.into());
            let value = coerce_input_value(builder, input_field.ty_record, field.value())?;
            fields_buffer.push((id, value));
            builder.value_path.pop();
        }
    } else {
        for input_field_id in input_object.input_field_ids {
            let input_field = &builder.graph[input_field_id];
            let name_id = input_field.name_id;
            let ty_record = input_field.ty_record;
            let default_value_id = input_field.default_value_id;

            builder.value_path.push(input_field.name_id.into());
            let value = if let Some(index) = fields.iter().position(|field| field.name() == builder[name_id]) {
                let field = fields.swap_remove(index);
                coerce_input_value(builder, ty_record, field.value())?
            } else if let Some(default_value_id) = default_value_id {
                builder.graph.input_values[default_value_id]
            } else if ty_record.wrapping.is_non_null() {
                return Err(InputValueError::UnexpectedNull {
                    expected: builder.type_name(ty_record),
                    path: builder.value_path_string(),
                });
            } else {
                builder.value_path.pop();
                continue;
            };

            fields_buffer.push((input_field_id, value));
            builder.value_path.pop();
        }
    }

    if let Some(field) = fields.first() {
        return Err(InputValueError::UnknownInputField {
            input_object: builder[builder.graph[input_object_id].name_id].clone(),
            name: field.name().to_owned(),
            path: builder.value_path_string(),
        });
    }

    // We iterate over input fields in order which is a range, so it should be sorted by the
    // id.
    debug_assert!(fields_buffer.is_sorted_by_key(|(id, _)| *id));
    let ids = builder.graph.input_values.append_input_object(&mut fields_buffer);
    builder.input_fields_buffer_pool.push(fields_buffer);
    Ok(SchemaInputValueRecord::InputObject(ids))
}

fn coerce_enum_value(
    builder: &mut GraphBuilder,
    enum_id: EnumDefinitionId,
    value: ConstValue,
) -> Result<SchemaInputValueRecord, InputValueError> {
    let r#enum = &builder.graph[enum_id];
    let value = match value {
        ConstValue::Enum(e) => e.as_str(),
        value => {
            return Err(InputValueError::IncorrectEnumValueType {
                r#enum: builder[r#enum.name_id].clone(),
                actual: value.into(),
                path: builder.value_path_string(),
            });
        }
    };

    for id in r#enum.value_ids {
        if builder[builder.graph[id].name_id] == value {
            return Ok(SchemaInputValueRecord::EnumValue(id));
        }
    }
    Err(InputValueError::UnknownEnumValue {
        r#enum: builder[r#enum.name_id].clone(),
        value: value.to_string(),
        path: builder.value_path_string(),
    })
}

fn coerce_scalar_value(
    builder: &mut GraphBuilder,
    scalar_id: ScalarDefinitionId,
    value: ConstValue,
) -> Result<SchemaInputValueRecord, InputValueError> {
    match builder.graph[scalar_id].ty {
        ScalarType::String => match value {
            ConstValue::String(s) => Some(builder.ingest_str(s.value())),
            _ => None,
        }
        .map(SchemaInputValueRecord::String),
        ScalarType::Float => match value {
            ConstValue::Int(n) => Some(n.value() as f64),
            ConstValue::Float(f) => Some(f.value()),
            _ => None,
        }
        .map(SchemaInputValueRecord::Float),
        ScalarType::Int => match value {
            ConstValue::Int(n) => {
                let n = i32::try_from(n.value()).map_err(|_| InputValueError::IncorrectScalarValue {
                    actual: n.to_string(),
                    expected: builder[builder.graph[scalar_id].name_id].clone(),
                    path: builder.value_path_string(),
                })?;
                Some(n)
            }
            ConstValue::Float(f) if can_coerce_to_int(f.value()) => Some(f.value() as i32),
            _ => None,
        }
        .map(SchemaInputValueRecord::Int),
        ScalarType::Boolean => match value {
            ConstValue::Boolean(b) => Some(b.value()),
            _ => None,
        }
        .map(SchemaInputValueRecord::Boolean),
        ScalarType::Unknown => {
            return Ok(ingest_arbitrary_value(builder, value));
        }
    }
    .ok_or_else(|| InputValueError::IncorrectScalarType {
        actual: value.into(),
        expected: builder[builder.graph[scalar_id].name_id].clone(),
        path: builder.value_path_string(),
    })
}

fn ingest_arbitrary_value(builder: &mut GraphBuilder, value: ConstValue) -> SchemaInputValueRecord {
    match value {
        ConstValue::Null(_) => SchemaInputValueRecord::Null,
        ConstValue::String(s) => SchemaInputValueRecord::String(builder.ingest_str(s.value())),
        ConstValue::Int(n) => SchemaInputValueRecord::I64(n.value()),
        ConstValue::Float(f) => SchemaInputValueRecord::Float(f.value()),
        ConstValue::Boolean(b) => SchemaInputValueRecord::Boolean(b.value()),
        ConstValue::Enum(s) => SchemaInputValueRecord::UnboundEnumValue(builder.ingest_str(s.as_str())),
        ConstValue::List(list) => {
            let ids = builder.graph.input_values.reserve_list(list.len());
            for (value, id) in list.into_iter().zip(ids) {
                builder.graph.input_values[id] = ingest_arbitrary_value(builder, value);
            }
            SchemaInputValueRecord::List(ids)
        }
        ConstValue::Object(fields) => {
            let ids = builder.graph.input_values.reserve_map(fields.len());
            for (field, id) in fields.into_iter().zip(ids) {
                let name = builder.ingest_str(field.name());
                let value = ingest_arbitrary_value(builder, field.value());
                builder.graph.input_values[id] = (name, value);
            }
            let ctx = &builder.ctx;
            builder.graph.input_values[ids]
                .sort_unstable_by(|(left_key, _), (right_key, _)| ctx[*left_key].cmp(&ctx[*right_key]));
            SchemaInputValueRecord::Map(ids)
        }
    }
}
