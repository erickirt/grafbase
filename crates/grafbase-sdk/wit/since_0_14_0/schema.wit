interface schema {
    record schema {
        // The IDs and this list have not relationship. They're not indices within this list.
        definitions: list<definition>,
    }

    variant definition {
        scalar(scalar-definition),
        object(object-definition),
        %interface(interface-definition),
        union(union-definition),
        %enum(enum-definition),
        input-object(input-object-definition),
    }

    record scalar-definition {
        id: u32,
        name: string,
        specified-by-url: option<string>,
        directives: list<directive>
    }

    record object-definition {
        id: u32,
        name: string,
        interfaces: list<u32>,
        fields: list<field-definition>,
        directives: list<directive>
    }

    record interface-definition {
        id: u32,
        name: string,
        interfaces: list<u32>,
        fields: list<field-definition>,
        directives: list<directive>
    }

    record union-definition {
        id: u32,
        name: string,
        member-types: list<u32>,
        directives: list<directive>
    }

    record enum-definition {
        id: u32,
        name: string,
        values: list<enum-value>,
        directives: list<directive>
    }

    record input-object-definition {
        id: u32,
        name: string,
        input-fields: list<input-value-definition>,
        directives: list<directive>
    }

    record field-definition {
        id: u32,
        name: string,
        ty: ty,
        arguments: list<input-value-definition>,
        directives: list<directive>
    }

    record ty {
        wrapping: list<wrapping-type>,
        definition-id: u32
    }

    enum wrapping-type {
        non-null,
        %list
    }

    record input-value-definition {
        id: u32,
        name: string,
        ty: ty,
        directives: list<directive>
    }

    record enum-value {
        name: string,
        directives: list<directive>
    }

    record directive {
        name: string,
        // encoded in CBOR
        arguments: list<u8>,
    }
}
