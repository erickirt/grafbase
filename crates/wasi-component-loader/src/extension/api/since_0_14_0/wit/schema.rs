use crate::state::WasiState;

pub use super::grafbase::sdk::schema::*;

impl Host for WasiState {}
