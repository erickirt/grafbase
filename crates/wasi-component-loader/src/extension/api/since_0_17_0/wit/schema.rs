pub use crate::extension::api::{
    since_0_9_0::wit::directive::{
        DirectiveSite, EnumDirectiveSite, FieldDefinitionDirective, FieldDefinitionDirectiveSite,
        InterfaceDirectiveSite, ObjectDirectiveSite, ScalarDirectiveSite, UnionDirectiveSite,
    },
    since_0_15_0::wit::schema::*,
};
