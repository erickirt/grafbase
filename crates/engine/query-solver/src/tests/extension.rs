use schema::Schema;

use crate::assert_solving_snapshots;

const SCHEMA: &str = r#"
directive @join__unionMember(graph: join__Graph!, member: String!) on UNION

directive @join__implements(graph: join__Graph!, interface: String!) on OBJECT | INTERFACE

directive @join__graph(name: String!, url: String) on ENUM_VALUE

directive @join__field(graph: join__Graph, requires: join__FieldSet, provides: join__FieldSet, type: String, external: Boolean, override: String, overrideLabel: String) on FIELD_DEFINITION | INPUT_FIELD_DEFINITION

directive @join__type(graph: join__Graph, key: join__FieldSet, extension: Boolean = false, resolvable: Boolean = true, isInterfaceObject: Boolean = false) on SCALAR | OBJECT | INTERFACE | UNION | ENUM | INPUT_OBJECT

directive @join__owner(graph: join__Graph!) on OBJECT

directive @composite__lookup on FIELD_DEFINITION

directive @composite__require(field: composite__FieldSelectionMap!) on ARGUMENT_DEFINITION

directive @composite__is(field: composite__FieldSelectionMap!) on FIELD_DEFINITION | ARGUMENT_DEFINITION

scalar DateTime

scalar Duration

scalar join__FieldSet

scalar composite__FieldSelectionMap

type Call
  @join__type(graph: ZENDESK)
{
  associated_deal_ids: [ID!]! @inaccessible
  deals: [Deal!]! @composite__derive(graph: ZENDESK) @composite__is(graph: ZENDESK, field: "associated_deal_ids[{ id: . }]")
  deals2(ids: [ID!]! @composite__require(graph: ZENDESK, field: "associated_deal_ids")): [Deal!]! @extension__directive(graph: ZENDESK, extension: REST, name: "rest", arguments: {endpoint: "zendesk", method: GET, path: "/deals?ids={{#args.ids}}{{.}}{{^-last}},{{/-last}}{{/args.ids}}", selection: "[.items[] | .data | {\n  id,\n  name,\n  createdAt: .created_at\n}]"})
  duration: Duration!
  id: ID!
  summary: String!
}

type Deal
  @join__type(graph: ZENDESK, key: "id", resolvable: false)
{
  createdAt: DateTime!
  id: ID!
  name: String!
  orders(id: ID! @composite__require(graph: ZENDESK, field: "id")): [Order!]! @extension__directive(graph: ZENDESK, extension: REST, name: "rest", arguments: {endpoint: "zendesk", method: GET, path: "/orders?deal_id={{args.id}}", selection: "[.items[] | .data | {\n  id,\n  name,\n  deal_id,\n  createdAt: .created_at\n}]"})
}

type Order
  @join__type(graph: ZENDESK)
{
  createdAt: DateTime!
  deal: Deal! @composite__derive(graph: ZENDESK)
  deal_id: ID! @inaccessible
  id: ID!
  lineItems(id: ID! @composite__require(graph: ZENDESK, field: "id")): [LineItem!]! @extension__directive(graph: ZENDESK, extension: REST, name: "rest", arguments: {endpoint: "zendesk", method: GET, path: "/orders/{{args.id}}/line_items", selection: "[.items[] | .data | {\n  id,\n  product_id,\n  quantity\n}]"})
}

type LineItem
  @join__type(graph: ZENDESK)
{
  id: ID!
  product: Product! @composite__derive(graph: ZENDESK)
  product_id: ID! @inaccessible
}

type Product
  @join__type(graph: ZENDESK, key: "id", resolvable: false)
{
  description: String!
  id: ID!
  name: String!
}

type Query
{
  calls: [Call!]! @extension__directive(graph: ZENDESK, extension: REST, name: "rest", arguments: {endpoint: "zendesk", method: GET, path: "/calls", selection: "[.items[] | .data | {\n  id,\n  summary,\n  duration,\n  associated_deal_ids\n}]"}) @join__field(graph: ZENDESK)
  dealLookup(id: ID!): Deal! @inaccessible @composite__lookup(graph: ZENDESK) @extension__directive(graph: ZENDESK, extension: REST, name: "rest", arguments: {endpoint: "zendesk", method: GET, path: "/deals/{{args.id}}", selection: ".data | {\n  id,\n  name,\n  createdAt: .created_at\n}"}) @join__field(graph: ZENDESK)
  productsLookup(ids: [ID!]!): [Product!]! @inaccessible @composite__lookup(graph: ZENDESK) @extension__directive(graph: ZENDESK, extension: REST, name: "rest", arguments: {endpoint: "zendesk", method: GET, path: "/products?ids={{#args.ids}}{{.}}{{^-last}},{{/-last}}{{/args.ids}}", selection: "[.items[] | .data | {\n  id,\n  name,\n  description\n}]"}) @join__field(graph: ZENDESK)
}

enum join__Graph
{
  ZENDESK @join__graph(name: "zendesk")
}

enum extension__Link
{
  REST @extension__link(url: "file:///rest/build", schemaDirectives: [{graph: ZENDESK, name: "restEndpoint", arguments: {name: "zendesk", baseURL: "http://localhost:8080/v2", headers: [{name: "Accept", value: "application/json"}]}}])
}
"#;

const REST_SDL: &str = r#"
scalar UrlTemplate
scalar JqTemplate

"""
@restEndpoint directive enables defining named REST endpoints for the
schema. The directive can be used multiple times on a schema to
define different endpoints.

Example:
extend schema
  @restEndpoint(name: "weather", baseURL: "https://api.weather.com")
  @restEndpoint(name: "users", baseURL: "https://api.users.example.com")
"""
directive @restEndpoint(
  """
  A unique identifier for the REST endpoint
  """
  name: String!
  """
  The base URL for the REST API
  """
  baseURL: String!
  "Header to send to the endpoint"
  headers: [HTTPHeaderMapping!]
) repeatable on SCHEMA

input HTTPHeaderMapping {
  "Name of the HTTP header"
  name: String!
  """
  Static header value. It's a template that accepts a `config` parameter which
  represents the TOML config associated with this extension. So for example:
  `value: "Bearer: {{ config.token }}"`
  """
  value: String!
}

"""
@rest directive allows you to define RESTful API integrations for GraphQL
fields. This directive maps GraphQL fields to REST endpoints, enabling
seamless integration between your GraphQL schema and external REST APIs.

Example:
type Query {
  users: [User] @rest(
    endpoint: "users",
    method: GET,
    path: "/users",
  )
}
"""
directive @rest(
  """
  The name of the REST endpoint to use, as defined by @restEndpoint
  """
  endpoint: String!

  """
  The HTTP method to use for the request, such as GET, POST, etc.
  """
  method: HttpMethod!

  """
  The path template for the request, which can include
  variable substitutions from GraphQL arguments.
  This supports templating using GraphQL arguments: {{args.myArgument}}
  """
  path: UrlTemplate!

  """
  Specifies which fields from the GraphQL selection to include in the
  response.
  """
  selection: JqTemplate

  """
  Configuration for the request body, can include static values or
  selections from the GraphQL arguments
  """
  body: Body = { selection: ".args.input" }
) on FIELD_DEFINITION

scalar JSON

"""
Body input type defines how to construct the request body for REST
API calls. It allows for dynamic construction from GraphQL arguments
or static values.
"""
input Body {
  """
  Specifies which GraphQL arguments to include in the request body.
  Use "*" to include all arguments, or provide specific field names.
  """
  selection: JqTemplate

  """
  Static JSON content to include in the request body,
  which will be merged with any selected values.
  """
  static: JSON
}

"""
HttpMethod enum represents the standard HTTP methods supported
for REST API interactions.
"""
enum HttpMethod {
  """
  HTTP GET method for retrieving resources
  """
  GET
  """
  HTTP POST method for creating resources
  """
  POST
  """
  HTTP PUT method for replacing resources
  """
  PUT
  """
  HTTP DELETE method for removing resources
  """
  DELETE
  """
  HTTP HEAD method for retrieving headers only
  """
  HEAD
  """
  HTTP OPTIONS method for describing communication options
  """
  OPTIONS
  """
  HTTP CONNECT method for establishing tunnels
  """
  CONNECT
  """
  HTTP TRACE method for diagnostic testing
  """
  TRACE
  """
  HTTP PATCH method for partial modifications
  """
  PATCH
}
"#;

#[tokio::test]
async fn mix_of_look_derive_require() {
    let tmpdir = tempfile::tempdir().unwrap();
    let manifest = extension_catalog::Manifest {
        id: "rest-1.0.0".parse().unwrap(),
        r#type: extension_catalog::Type::Resolver(Default::default()),
        sdk_version: "0.0.0".parse().unwrap(),
        minimum_gateway_version: "0.0.0".parse().unwrap(),
        description: String::new(),
        sdl: Some(REST_SDL.into()),
        readme: None,
        homepage_url: None,
        repository_url: None,
        license: None,
        permissions: Default::default(),
        legacy_event_filter: Default::default(),
    };

    std::fs::write(
        tmpdir.path().join("manifest.json"),
        serde_json::to_vec(&manifest.clone().into_versioned()).unwrap(),
    )
    .unwrap();

    let mut catalog = extension_catalog::ExtensionCatalog::default();
    let wasm_path = tmpdir.path().join("extension.wasm");
    std::fs::write(&wasm_path, b"wasm").unwrap();
    catalog.push(extension_catalog::Extension {
        config_key: String::new(),
        manifest,
        wasm_path,
    });

    let schema = Schema::builder(&SCHEMA.replace(
        "file:///rest/build",
        url::Url::from_file_path(tmpdir.path()).unwrap().as_str(),
    ))
    .extensions(None, &catalog)
    .build()
    .await
    .unwrap();

    // Extension resolver can be placed on arbitrary fields, its presence indicates that it must be
    // used to resolve the field. We're taking this into account when computing the providable
    // fields. However this logic was initially only applied for nested fields. Not the root
    // fields, which isn't necessary except for lookup. While building the schema we don't apply
    // the same, complex, logic. Everything that isn't part of the key will be assigned a lookup
    // resolver. So it's during the query planning that we need to check whether a field is
    // actually providable or not.
    assert_solving_snapshots!(
        "mix_of_look_derive_require",
        schema,
        r#"
        query Calls {
          calls {
            deals {
              orders {
                createdAt
                lineItems {
                  product {
                    name
                    description
                  }
                  id
                }
                id
              }
              createdAt
            }
          }
        }
        "#
    );
}
