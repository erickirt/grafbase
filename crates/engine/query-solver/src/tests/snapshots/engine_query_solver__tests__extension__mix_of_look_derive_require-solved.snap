---
source: crates/engine/query-solver/src/tests/extension.rs
expression: "digraph {\n    0 [ label = \"root\", color=forestgreen ]\n    1 [ label = \"Deal.createdAt\", color=forestgreen ]\n    2 [ label = \"Order.createdAt\", color=forestgreen ]\n    3 [ label = \"Order.id\", color=forestgreen ]\n    4 [ label = \"LineItem.id\", color=forestgreen ]\n    5 [ label = \"Product.name\", color=forestgreen ]\n    6 [ label = \"Product.description\", color=forestgreen ]\n    7 [ label = \"rest/rest-1.0.0#zendesk\", shape=parallelogram, color=dodgerblue, color=forestgreen ]\n    8 [ label = \"calls#zendesk\", shape=box, color=dodgerblue, color=forestgreen ]\n    9 [ label = \"deals#zendesk@derive\", shape=box, color=dodgerblue, color=forestgreen ]\n    10 [ label = \"rest/rest-1.0.0#zendesk\", shape=parallelogram, color=dodgerblue, color=forestgreen ]\n    11 [ label = \"orders#zendesk\", shape=box, color=dodgerblue, color=forestgreen ]\n    12 [ label = \"createdAt#zendesk\", shape=box, color=dodgerblue, color=forestgreen ]\n    13 [ label = \"rest/rest-1.0.0#zendesk\", shape=parallelogram, color=dodgerblue, color=forestgreen ]\n    14 [ label = \"lineItems#zendesk\", shape=box, color=dodgerblue, color=forestgreen ]\n    15 [ label = \"product#zendesk@derive\", shape=box, color=dodgerblue, color=forestgreen ]\n    16 [ label = \"Lookup/rest/rest-1.0.0#zendesk\", shape=parallelogram, color=dodgerblue, color=forestgreen ]\n    17 [ label = \"name#zendesk\", shape=box, color=dodgerblue, color=forestgreen ]\n    18 [ label = \"description#zendesk\", shape=box, color=dodgerblue, color=forestgreen ]\n    19 [ label = \"id#zendesk\", shape=box, color=dodgerblue, color=forestgreen ]\n    20 [ label = \"id#zendesk\", shape=box, color=dodgerblue, color=forestgreen ]\n    21 [ label = \"Lookup/rest/rest-1.0.0#zendesk\", shape=parallelogram, color=dodgerblue, color=forestgreen ]\n    22 [ label = \"createdAt#zendesk\", shape=box, color=dodgerblue, color=forestgreen ]\n    23 [ label = \"*Deal.id\", color=forestgreen ]\n    24 [ label = \"id#zendesk@derive\", shape=box, color=dodgerblue, color=forestgreen ]\n    25 [ label = \"*Product.id\", color=forestgreen ]\n    26 [ label = \"id#zendesk@derive\", shape=box, color=dodgerblue, color=forestgreen ]\n    27 [ label=\"\", style=dashed]\n    0 -> 7 [ label = \"\", color=forestgreen,fontcolor=forestgreen ]\n    7 -> 8 [ label = \"\", color=forestgreen,fontcolor=forestgreen ]\n    8 -> 9 [ label = \"\", color=forestgreen,fontcolor=forestgreen ]\n    9 -> 10 [ label = \"\", color=forestgreen,fontcolor=forestgreen ]\n    10 -> 11 [ label = \"\", color=forestgreen,fontcolor=forestgreen ]\n    11 -> 12 [ label = \"\", color=forestgreen,fontcolor=forestgreen ]\n    12 -> 2 [ label = \"\", color=forestgreen,fontcolor=forestgreen ]\n    11 -> 13 [ label = \"\", color=forestgreen,fontcolor=forestgreen ]\n    13 -> 14 [ label = \"\", color=forestgreen,fontcolor=forestgreen ]\n    14 -> 15 [ label = \"\", color=forestgreen,fontcolor=forestgreen ]\n    15 -> 16 [ label = \"\", color=forestgreen,fontcolor=forestgreen ]\n    16 -> 17 [ label = \"\", color=forestgreen,fontcolor=forestgreen ]\n    17 -> 5 [ label = \"\", color=forestgreen,fontcolor=forestgreen ]\n    16 -> 18 [ label = \"\", color=forestgreen,fontcolor=forestgreen ]\n    18 -> 6 [ label = \"\", color=forestgreen,fontcolor=forestgreen ]\n    14 -> 19 [ label = \"\", color=forestgreen,fontcolor=forestgreen ]\n    19 -> 4 [ label = \"\", color=forestgreen,fontcolor=forestgreen ]\n    11 -> 20 [ label = \"\", color=forestgreen,fontcolor=forestgreen ]\n    20 -> 3 [ label = \"\", color=forestgreen,fontcolor=forestgreen ]\n    9 -> 21 [ label = \"\", color=forestgreen,fontcolor=forestgreen ]\n    21 -> 22 [ label = \"\", color=forestgreen,fontcolor=forestgreen ]\n    22 -> 1 [ label = \"\", color=forestgreen,fontcolor=forestgreen ]\n    9 -> 24 [ label = \"\", color=forestgreen,fontcolor=forestgreen ]\n    24 -> 23 [ label = \"\", color=forestgreen,fontcolor=forestgreen ]\n    15 -> 26 [ label = \"\", color=forestgreen,fontcolor=forestgreen ]\n    26 -> 25 [ label = \"\", color=forestgreen,fontcolor=forestgreen ]\n    27 -> 0 [ label = \"\", color=royalblue,fontcolor=royalblue,style=dashed ]\n}\n"
---
digraph {
    0 [ label = "root", steiner=1 ]
    1 [ label = "Deal.createdAt", steiner=1 ]
    2 [ label = "Order.createdAt", steiner=1 ]
    3 [ label = "Order.id", steiner=1 ]
    4 [ label = "LineItem.id", steiner=1 ]
    5 [ label = "Product.name", steiner=1 ]
    6 [ label = "Product.description", steiner=1 ]
    7 [ label = "rest/rest-1.0.0#zendesk", steiner=1 ]
    8 [ label = "calls#zendesk", steiner=1 ]
    9 [ label = "deals#zendesk@derive", steiner=1 ]
    10 [ label = "rest/rest-1.0.0#zendesk", steiner=1 ]
    11 [ label = "orders#zendesk", steiner=1 ]
    12 [ label = "createdAt#zendesk", steiner=1 ]
    13 [ label = "rest/rest-1.0.0#zendesk", steiner=1 ]
    14 [ label = "lineItems#zendesk", steiner=1 ]
    15 [ label = "product#zendesk@derive", steiner=1 ]
    16 [ label = "Lookup/rest/rest-1.0.0#zendesk", steiner=1 ]
    17 [ label = "name#zendesk", steiner=1 ]
    18 [ label = "description#zendesk", steiner=1 ]
    19 [ label = "id#zendesk", steiner=1 ]
    20 [ label = "id#zendesk", steiner=1 ]
    21 [ label = "Lookup/rest/rest-1.0.0#zendesk", steiner=1 ]
    22 [ label = "createdAt#zendesk", steiner=1 ]
    23 [ label = "*Deal.id", steiner=1 ]
    24 [ label = "id#zendesk@derive", steiner=1 ]
    25 [ label = "*Product.id", steiner=1 ]
    26 [ label = "id#zendesk@derive", steiner=1 ]
    27 [ label="", style=dashed]
    0 -> 7 [ cost=0, steiner=1]
    7 -> 8 [ cost=0, steiner=1]
    8 -> 9 [ cost=0, steiner=1]
    9 -> 10 [ cost=0, steiner=1]
    10 -> 11 [ cost=0, steiner=1]
    11 -> 12 [ cost=0, steiner=1]
    12 -> 2 [ cost=0, steiner=1]
    11 -> 13 [ cost=0, steiner=1]
    13 -> 14 [ cost=0, steiner=1]
    14 -> 15 [ cost=0, steiner=1]
    15 -> 16 [ cost=0, steiner=1]
    16 -> 17 [ cost=0, steiner=1]
    17 -> 5 [ cost=0, steiner=1]
    16 -> 18 [ cost=0, steiner=1]
    18 -> 6 [ cost=0, steiner=1]
    14 -> 19 [ cost=0, steiner=1]
    19 -> 4 [ cost=0, steiner=1]
    11 -> 20 [ cost=0, steiner=1]
    20 -> 3 [ cost=0, steiner=1]
    9 -> 21 [ cost=0, steiner=1]
    21 -> 22 [ cost=0, steiner=1]
    22 -> 1 [ cost=0, steiner=1]
    9 -> 24 [ cost=0, steiner=1]
    24 -> 23 [ cost=0, steiner=1]
    15 -> 26 [ cost=0, steiner=1]
    26 -> 25 [ cost=0, steiner=1]
    27 -> 0 [ cost=0, steiner=0]
}
